features/keychain/module/multi-seed-keychain.js

```diff
@@ -0,0 +1,141 @@
+import assert from 'minimalistic-assert'
+import { Keychain } from './keychain'
+import { getSeedId } from './crypto/seed-id'
+
+const MODULE_ID = 'multiSeedKeychain'
+
+class MultiSeedKeychain {
+  #keychains = Object.create(null)
+  #primarySeedId
+  #legacyPrivToPub
+  #logger
+  constructor({ legacyPrivToPub = Object.create(null), logger }) {
+    this.#legacyPrivToPub = legacyPrivToPub
+    this.#logger = logger
+    this.sodium = new Proxy(
+      {},
+      {
+        get:
+          (target, key) =>
+          async ({ walletAccount, ...rest }) =>
+            this.#getKeychainForWalletAccount(walletAccount).sodium[key](rest),
+      }
+    )
+
+    this.secp256k1 = new Proxy(
+      {},
+      {
+        get:
+          (target, key) =>
+          async ({ walletAccount, ...rest }) =>
+            this.#getKeychainForWalletAccount(walletAccount).secp256k1[key](rest),
+      }
+    )
+
+    this.ed25519 = new Proxy(
+      {},
+      {
+        get:
+          (target, key) =>
+          async ({ walletAccount, ...rest }) =>
+            this.#getKeychainForWalletAccount(walletAccount).ed25519[key](rest),
+      }
+    )
+  }
+
+  #getKeychainForWalletAccount = (walletAccount) => {
+    assert(this.#keychains, 'keychain is locked')
+
+    const seedId = walletAccount.source === 'exodus' ? this.#primarySeedId : walletAccount.id
+    assert(seedId, `walletAccount has no seed id: ${seedId}`)
+    assert(this.#keychains[seedId], `keychain not found for seed id: ${seedId}`)
+    return this.#keychains[seedId]
+  }
+
+  lock = () => {
+    this.#keychains = null
+    this.#primarySeedId = null
+  }
+
+  #initKeychain = (seed) => {
+    if (!this.#keychains) this.#keychains = Obejct.create(null)
+
+    const seedId = getSeedId(seed)
+    if (this.#keychains[seedId]) throw new Error(`already have seed with id: ${seedId}`)
+
+    const keychain = new Keychain({
+      legacyPrivToPub: this.#legacyPrivToPub,
+      logger: this.#logger,
+    })
+
+    keychain.unlock({ seed })
+
+    this.#keychains[seedId] = keychain
+    return seedId
+  }
+
+  setPrimarySeed = (seed) => {
+    this.#primarySeedId = this.#initKeychain(seed)
+    return this.#primarySeedId
+  }
+
+  addSeed = (seed) => {
+    assert(this.#keychains, 'keychain is locked')
+    assert(this.#primarySeedId, 'primary seed not set')
+    return this.#initKeychain(seed)
+  }
+
+  // unlock({ seeds, primarySeed }) {
```

Comment:
not sure unlock makes sense as a concept for multiSeedKeychain, because we'd need to provide all seeds on unlock. Wdyt about `setPrimarySeed`/`addSeed` + `lock`?

Reply:
Sgtm. I liked the symmetry of `unlock/lock`. Maybe lock and unlock can be per seed basis? Like both will accept a seed identifier as argument

Reply:
the wording is not ideal imo. What `unlock` currently does is initializing each KDF with the seed. The actual "unlocking" happens in the wallet module. I would prefer `load({ seeds })` and maybe `unload` to distinguish it from `clear` which we run on restore/delete, but maybe `clear` is also fine. 

I would also be happier if we could remove the notion of primary/other seeds from the keychain entirely.

Reply:
not all seeds are available at all times. You'd still need an addSeed when the user imports a new seed right?

Reply:
getting rid of "primary" would be good though. How do we know which one's primary though, for things like seed derived ids, fusion keys, etc?

Reply:
haha for fusion we currently use [wallet.seed](https://github.com/exodusmovement/exodus-mobile/blob/02e4b42960adbb9c2ef8ad5a190a90c3f54b048d/src/background/features/fusion/index.js#L73), another field i was hoping to kill

Reply:
> or would load "add" seeds instead of replace

I'd say it should be additive

> getting rid of "primary" would be good though. How do we know which one's primary though, for things like seed derived ids, fusion keys, etc?

wallet manages the seed, so maybe we could have an abstraction on top of keychain and wallet which mirrors the keychain operations but always passes the primary seed id? another option would be to expose a getter for the primary seed id on wallet and let consumers collect get and supply that themselves

Reply:
prob a primarySeedIdAtom would do the trick, though we might see it everywhere

Reply:
addSeed seems more explicitly additive, but i don't feel too strongly about this decision, lmk

i'm not sure what clear's meant to do today. Should it lock @diegomura ?

Reply:
> addSeed seems more explicitly additive

fair, let's keep